src/dmd/dtoh.d

```diff
@@ -966,100 +1034,165 @@ public:
-        if (ed.ident)
-            ident = ed.ident.toChars();
-        if (!ident)
-        {
-            buf.writestring("enum");
-        }
-        else if (hasBaseType)
+        // determine if this is an enum, or just a group of manifest constants
+        bool manifestConstants = !type || (isAnonymous && kind == EnumKind.Other);
+        assert(!manifestConstants || isAnonymous);
+
+        const(char)[] name = null;
+        char[] nameLower = null;
+        char[] nameUpper = null;
+        char[] nameAbbrev = null;
+        if (!isAnonymous)
         {
-            //printf("typedef _d_%s %s;\n", ed.memtype.kind, ident);
-            if (global.params.cplusplus >= CppStdRevision.cpp11)
+            name = ed.ident.toString;
+            nameLower = new char[name.length];
+            nameUpper = new char[name.length];
+            foreach (i, c; name)
             {
-                //printf("Using cpp 11 and beyond\n");
-                buf.printf("enum %s : %s", ident, ed.memtype.kind);
+                nameUpper[i] = cast(char)toupper(c);
+                nameLower[i] = cast(char)tolower(c);
+                if (isupper(c))
+                    nameAbbrev ~= c;
             }
+        }
+
+        // write the enum header
+        if (!manifestConstants)
+        {
+            indent();
+            if (!isAnonymous && kind == EnumKind.Int)
+                buf.writestring("BEGIN_ENUM(");
+            else if (!isAnonymous && kind == EnumKind.Numeric)
+                buf.writestring("BEGIN_ENUM_NUMERIC(");
+            else if(!isAnonymous)
+                buf.writestring("BEGIN_ENUM_TYPE(");
+            else if (kind == EnumKind.Int)
+                buf.writestring("BEGIN_ANON_ENUM(");
             else
+                buf.writestring("BEGIN_ANON_ENUM_NUMERIC(");
+            if (kind != EnumKind.Int)
+                writeEnumTypeName(type);
+            if (!isAnonymous)
             {
-                //printf("Using cpp 98\n");
-                buf.writestring("typedef ");
-                buf.writestring(translateBasicType(ed.memtype.ty));
-                buf.writeByte(' ');
-                buf.writestring(ident);
-                buf.writestring(";\n");
-                buf.writestring("enum");
+                if (kind != EnumKind.Int)
+                    buf.writestring(", ");
+                buf.writestring(name);
+                buf.writestring(", ");
+                buf.writestring(nameUpper);
+                buf.writestring(", ");
+                buf.writestring(nameLower);
             }
-        }
-        else
-        {
-            buf.writestring("enum ");
-            buf.writestring(ident);
+            buf.writestring(")\n");
         }
 
-        if (!ed.members)
+        // emit constant for each member
+        foreach (i, _m; *ed.members)
         {
-            buf.writestring(";\n\n");
-            return;
+            AST.EnumMember m = cast(AST.EnumMember)_m;
```

Comment:
bad cast, bad, :-)

Reply:
Are other kinds of members possible in an `EnumDeclaration`?